tests/agent_server/test.py

```diff
@@ -0,0 +1,293 @@
+    )
+
+    last_change_3 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    first_change_1 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key1",
+    )
+
+    first_change_2 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key2",
+    )
+
+    first_change_3 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    import json
+    logger.info("Last change for key1: " + json.dumps(last_change_1, indent=2))
+    logger.info("Last change for key2: " + json.dumps(last_change_2, indent=2))
+    logger.info("Last change for key3: " + json.dumps(last_change_3, indent=2))
+    logger.info("First change for key1: " + json.dumps(first_change_1, indent=2))
+    logger.info("First change for key2: " + json.dumps(first_change_2, indent=2))
+    logger.info("First change for key3: " + json.dumps(first_change_3, indent=2))
+
+
+    async def get_dependencies(id: str) -> str:
+        logger.info(id)
+        result = await client.get_resource(environment, id=id, logs=True)
+        assert result.code == 200
+
+        return result.result.get("resource", {}).get("attributes", {}).get("requires")
+
+
+    async def need_redeploy(
+        resource_type: str,
+        agent: str,
+        key: str,
+        value: str,
+    ) -> bool:
+        last_deployment = await get_last_change(resource_type, agent, key, value)
+        logger.info(f"Last deployment of self: {last_deployment['finished']}")
+
+        # Check dependencies
+        dependencies = await get_dependencies(
+            f"{resource_type}[{agent},{key}={value}],v={last_deployment.get('version')}"
+        )
+
+        resource_id = re.compile(r"(.*)\[(.*),(.*)=(.*)\]")
+
+        # For each dependency, check first change after last_deployment
+        # If any change, needs redeploy
+        for dependency in dependencies:
+            reg_match = resource_id.match(dependency)
+            dep_resource_type = reg_match.group(1)
+            dep_agent = reg_match.group(2)
+            dep_key = reg_match.group(3)
+            dep_value = reg_match.group(4)
+
+            logger.info(f"Checking dependency {dependency}")
+            first_change = await get_first_change(
+                resource_type=dep_resource_type,
+                agent=dep_agent,
+                key=dep_key,
+                value=dep_value,
+                after=last_deployment["finished"],
+            )
+            if first_change is not None:
+                logger.info(f"Last redeployment of dependency: {first_change['finished']} (version {first_change['version']})")
+                return True
+
+        return False
+    
+    logger.info(await need_redeploy(
```

Comment:
The testcase doesn't cover most scenarios:

1. what with incremental deploys   (easy to test by triggering a repair run or making of the deploys an incremental)
2. what with partial deploys  (bit trickier to test)
3. what with repairs  (easy to test by using the provider to set a value, then trigger a repair run)
4. what with multiple deploys on the same version (can be done with 1 or 3)

Reply:
5. dryruns in between

Reply:
6. what for the first deploy of the resource
7. what for the first deploy of a dependant resource

Reply:
8. Test access right for agent client (enable authentication)